File: test/contracts.test.ts
import test from 'node:test';
import assert from 'node:assert/strict';
import { readdirSync, readFileSync, existsSync } from 'node:fs';
import { join, resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { Ajv2020 } from 'ajv/dist/2020.js';

import {
  DOCUMENT_SCHEMAS,
  ContractError,
  canonicalJson,
  inputFingerprint,
  parseAnyDocument,
  parseDocument,
  sha256Hex,
  type DocumentSchemaName,
} from '../ts/src/index.js';

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const FIXTURES = join(ROOT, 'fixtures');
const SCHEMAS = join(ROOT, 'schemas');

const readJson = (path: string): unknown => JSON.parse(readFileSync(path, 'utf8'));

// Fixture directory → the `schema` value its documents carry.
const FIXTURE_DOCS: Record<string, DocumentSchemaName> = {
  'bar-binary-layout.v1': 'heisentick/bar-binary-layout',
  'candle-snapshot-manifest.v1': 'heisentick/candle-snapshot-manifest',
  'validation-run-manifest.v1': 'heisentick/validation-run-manifest',
  'validation-run-request.v1': 'heisentick/validation-run-request',
  'validation-run-result.v1': 'heisentick/validation-run-result',
  'validation-run-result-failed.v1': 'heisentick/validation-run-result',
  'validation-request-message.v1': 'heisentick/validation-request-message',
  'validation-result-message.v1': 'heisentick/validation-result-message',
};

// The JSON Schema files themselves, compiled with Ajv, are the reference the
// generated zod must agree with on every fixture. Ajv is a dev dependency
// only; consumers get zod.
function ajvValidators() {
  const ajv = new Ajv2020({ strict: true, allowUnionTypes: true, allErrors: true });
  for (const file of readdirSync(SCHEMAS).filter((f) => f.endsWith('.schema.json'))) {
    ajv.addSchema(readJson(join(SCHEMAS, file)) as object, file);
  }
  return {
    validate(schemaFile: string, ref: string | null, value: unknown): boolean {
      const key = ref ? `${schemaFile}#${ref}` : schemaFile;
      const fn = ajv.getSchema(key);
      assert.ok(fn, `ajv could not resolve ${key}`);
      const ok = Boolean(fn(value));
      if (!ok && process.env.AJV_DEBUG) console.error(key, JSON.stringify(fn.errors));
      return ok;
    },
  };
}

const CROSS_FIELD_FIXTURES = new Set(['succeeded-without-report.json', 'succeeded-with-null-headline.json']);

const ajvRefFor: Record<string, [string, string | null]> = {
  'bar-binary-layout.v1': ['bar-binary-layout.v1.schema.json', null],
  'candle-snapshot-manifest.v1': ['candle-snapshot-manifest.v1.schema.json', null],
  'validation-run-manifest.v1': ['validation-run-manifest.v1.schema.json', null],
  'validation-run-request.v1': ['validation-run-request.v1.schema.json', null],
  'validation-run-result.v1': ['validation-run-result.v1.schema.json', null],
  'validation-run-result-failed.v1': ['validation-run-result.v1.schema.json', null],
  'validation-request-message.v1': ['validation-queue-messages.v1.schema.json', '/$defs/validationRequestMessage'],
  'validation-result-message.v1': ['validation-queue-messages.v1.schema.json', '/$defs/validationResultMessage'],
};

test('every fixture directory is covered and every document has both valid and invalid fixtures', () => {
  const dirs = readdirSync(FIXTURES).filter((d) => d !== 'canonical').sort();
  assert.deepEqual(dirs, Object.keys(FIXTURE_DOCS).sort());
  for (const dir of dirs) {
    assert.ok(readdirSync(join(FIXTURES, dir, 'valid')).length >= 1, `${dir} has a valid fixture`);
    assert.ok(readdirSync(join(FIXTURES, dir, 'invalid')).length >= 3, `${dir} has invalid fixtures`);
  }
});

test('zod and JSON Schema agree: valid fixtures pass both, invalid fixtures fail both', () => {
  const ajv = ajvValidators();
  for (const [dir, name] of Object.entries(FIXTURE_DOCS)) {
    const [schemaFile, ref] = ajvRefFor[dir];
    for (const file of readdirSync(join(FIXTURES, dir, 'valid'))) {
      const value = readJson(join(FIXTURES, dir, 'valid', file));
      assert.equal(ajv.validate(schemaFile, ref, value), true, `${dir}/valid/${file} (json schema)`);
      assert.doesNotThrow(() => parseDocument(name, value), `${dir}/valid/${file} (zod)`);
      assert.equal(parseAnyDocument(value).schema, name);
    }
    for (const file of readdirSync(join(FIXTURES, dir, 'invalid'))) {
      const value = readJson(join(FIXTURES, dir, 'invalid', file));
      // Cross-field rules live in code, not in the schema (see crossFieldIssues).
      if (!CROSS_FIELD_FIXTURES.has(file)) {
        assert.equal(ajv.validate(schemaFile, ref, value), false, `${dir}/invalid/${file} should fail json schema`);
      }
      assert.throws(() => parseDocument(name, value), ContractError, `${dir}/invalid/${file} should fail the package`);
    }
  }
});

test('unknown fields are rejected on every document', () => {
  for (const [name, schema] of Object.entries(DOCUMENT_SCHEMAS)) {
    const dir = Object.entries(FIXTURE_DOCS).find(([, n]) => n === name)?.[0];
    assert.ok(dir, `fixture dir for ${name}`);
    const value = readJson(join(FIXTURES, dir, 'valid', 'example.json')) as Record<string, unknown>;
    assert.equal(schema.safeParse({ ...value, unexpectedField: 1 }).success, false, name);
  }
});

test('an unknown or missing schema name is a ContractError, never a pass', () => {
  assert.throws(() => parseAnyDocument({ schema: 'heisentick/nope', version: 1 }), /unknown document schema "heisentick\/nope"/);
  assert.throws(() => parseAnyDocument({ version: 1 }), /missing schema field/);
  assert.throws(() => parseAnyDocument(null), ContractError);
});

test('canonical JSON matches the committed cases and rejects non-finite numbers and undefined', () => {
  const cases = readJson(join(FIXTURES, 'canonical', 'cases.json')) as { name: string; input: unknown; canonical: string; sha256?: string }[];
  for (const c of cases) {
    assert.equal(canonicalJson(c.input), c.canonical, c.name);
    if (c.sha256) assert.equal(sha256Hex(c.canonical), c.sha256, `${c.name} sha256`);
  }
  assert.throws(() => canonicalJson({ a: Number.NaN }), /non-finite number at \.a/);
  assert.throws(() => canonicalJson({ a: [1, Infinity] }), /non-finite number at \.a\[1\]/);
  assert.throws(() => canonicalJson({ a: undefined }), /undefined value at \.a/);
  assert.equal(canonicalJson({ b: -0 }), '{"b":0}');
});

test('bounded generated canonical cases preserve cross-language vectors', () => {
  const fixture = readJson(join(FIXTURES, 'canonical', 'generated-cases.json')) as {
    algorithm: string;
    boundedCaseCount: number;
    cases: { name: string; seed: string; input: unknown; canonical: string; sha256: string }[];
    fingerprints: unknown[];
  };
  assert.equal(fixture.algorithm, 'lcg32-v1');
  assert.ok(fixture.cases.length > 0, 'generated fixture has no canonical cases');
  assert.ok(fixture.fingerprints.length > 0, 'generated fixture has no fingerprint cases');
  assert.equal(fixture.cases.length, fixture.boundedCaseCount);
  for (const c of fixture.cases) {
    assert.equal(canonicalJson(c.input), c.canonical, c.name + ' seed=' + c.seed);
    assert.equal(sha256Hex(c.canonical), c.sha256, c.name + ' seed=' + c.seed + ' sha256');
  }
});

test('generated fingerprints ignore identity and detect material changes', () => {
  const fixture = readJson(join(FIXTURES, 'canonical', 'generated-cases.json')) as {
    fingerprints: {
      name: string;
      seed: string;
      manifest: Record<string, unknown>;
      identityVariant: Record<string, unknown>;
      materialVariant: { costs: Record<string, unknown> };
      fingerprint: string;
      materialFingerprint: string;
    }[];
  };
  for (const c of fixture.fingerprints) {
    const got = inputFingerprint(c.manifest);
    assert.equal(got, c.fingerprint, c.name + ' seed=' + c.seed);
    assert.equal(inputFingerprint({ ...c.manifest, ...c.identityVariant }), got, c.name + ' identity seed=' + c.seed);
    const material = {
      ...c.manifest,
      costs: { ...(c.manifest.costs as Record<string, unknown>), ...c.materialVariant.costs },
    };
    const materialFingerprint = inputFingerprint(material);
    assert.equal(materialFingerprint, c.materialFingerprint, c.name + ' material seed=' + c.seed);
    assert.notEqual(materialFingerprint, got, c.name + ' material changed seed=' + c.seed);
  }
});

test('inputFingerprint ignores submission identity and changes with any input', () => {
  const manifest = readJson(join(FIXTURES, 'validation-run-manifest.v1', 'valid', 'example.json')) as Record<string, unknown>;
  const fp = inputFingerprint(manifest);
  assert.match(fp, /^[0-9a-f]{64}$/);
  assert.equal(inputFingerprint({ ...manifest, runId: '9f3c1d2e-4a5b-4c6d-8e9f-0a1b2c3d4e5f', submittedAt: 1, requestedBy: { subject: 'x', via: 'local' } }), fp);
  assert.notEqual(inputFingerprint({ ...manifest, costs: { ...(manifest.costs as object), slippage: 0.07 } }), fp);
  const expectedPath = join(FIXTURES, 'validation-run-manifest.v1', 'fingerprint.txt');
  if (existsSync(expectedPath)) assert.equal(fp, readFileSync(expectedPath, 'utf8').trim());
});

File: contracts_test.go
package contracts

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/spik3r/heisentick-contracts/gen"
)

// Fixture directory → document schema name. Mirrors test/contracts.test.ts so
// both languages judge the same files.
var fixtureDocs = map[string]string{
	"bar-binary-layout.v1":            "heisentick/bar-binary-layout",
	"candle-snapshot-manifest.v1":     "heisentick/candle-snapshot-manifest",
	"validation-run-manifest.v1":      "heisentick/validation-run-manifest",
	"validation-run-request.v1":       "heisentick/validation-run-request",
	"validation-run-result.v1":        "heisentick/validation-run-result",
	"validation-run-result-failed.v1": "heisentick/validation-run-result",
	"validation-request-message.v1":   "heisentick/validation-request-message",
	"validation-result-message.v1":    "heisentick/validation-result-message",
}

func readFixtures(t *testing.T, dir, kind string) map[string][]byte {
	t.Helper()
	entries, err := os.ReadDir(filepath.Join("fixtures", dir, kind))
	if err != nil {
		t.Fatalf("read %s/%s: %v", dir, kind, err)
	}
	out := map[string][]byte{}
	for _, e := range entries {
		raw, err := os.ReadFile(filepath.Join("fixtures", dir, kind, e.Name()))
		if err != nil {
			t.Fatal(err)
		}
		out[e.Name()] = raw
	}
	return out
}

func TestFixturesValidateLikeTypeScript(t *testing.T) {
	for dir, name := range fixtureDocs {
		for file, raw := range readFixtures(t, dir, "valid") {
			if _, err := Validate(name, raw); err != nil {
				t.Errorf("%s/valid/%s: %v", dir, file, err)
			}
			gotName, _, err := ValidateAny(raw)
			if err != nil || gotName != name {
				t.Errorf("%s/valid/%s: ValidateAny → %q, %v", dir, file, gotName, err)
			}
		}
		for file, raw := range readFixtures(t, dir, "invalid") {
			_, err := Validate(name, raw)
			if err == nil {
				t.Errorf("%s/invalid/%s: expected rejection", dir, file)
				continue
			}
			if _, ok := err.(*ContractError); !ok {
				t.Errorf("%s/invalid/%s: expected ContractError, got %T", dir, file, err)
			}
		}
	}
}

func TestUnknownSchemaIsAnError(t *testing.T) {
	if _, err := Validate("heisentick/nope", []byte(`{}`)); err == nil || !strings.Contains(err.Error(), "unknown document schema") {
		t.Fatalf("expected unknown schema error, got %v", err)
	}
	if _, _, err := ValidateAny([]byte(`{"version":1}`)); err == nil || !strings.Contains(err.Error(), "missing schema field") {
		t.Fatalf("expected missing schema error, got %v", err)
	}
}

func TestDecodeIntoGeneratedStruct(t *testing.T) {
	raw, err := os.ReadFile("fixtures/candle-snapshot-manifest.v1/valid/example.json")
	if err != nil {
		t.Fatal(err)
	}
	var m gen.CandleSnapshotManifestV1
	if err := Decode("heisentick/candle-snapshot-manifest", raw, &m); err != nil {
		t.Fatal(err)
	}
	if m.Symbol != "XAUUSD" || m.Count != 2 || string(m.Timeframe) != "1m" {
		t.Fatalf("unexpected decode: %+v", m)
	}
	bad := strings.Replace(string(raw), `"timeframe": "1m"`, `"timeframe": "2h"`, 1)
	if err := Decode("heisentick/candle-snapshot-manifest", []byte(bad), &m); err == nil {
		t.Fatal("Decode must validate before unmarshalling")
	}
}

func TestCanonicalCasesMatchTypeScript(t *testing.T) {
	raw, err := os.ReadFile("fixtures/canonical/cases.json")
	if err != nil {
		t.Fatal(err)
	}
	var cases []struct {
		Name      string          `json:"name"`
		Input     json.RawMessage `json:"input"`
		Canonical string          `json:"canonical"`
		SHA256    string          `json:"sha256"`
	}
	if err := json.Unmarshal(raw, &cases); err != nil {
		t.Fatal(err)
	}
	if len(cases) == 0 {
		t.Fatal("no canonical cases")
	}
	for _, c := range cases {
		got, err := CanonicalJSONBytes(c.Input)
		if err != nil {
			t.Errorf("%s: %v", c.Name, err)
			continue
		}
		if got != c.Canonical {
			t.Errorf("%s:\n got %s\nwant %s", c.Name, got, c.Canonical)
		}
		if c.SHA256 == "" {
			t.Errorf("%s: fixture not sealed (run scripts/seal-fixtures.mjs)", c.Name)
		} else if SHA256Hex(got) != c.SHA256 {
			t.Errorf("%s: sha256 mismatch", c.Name)
		}
	}
}

func TestGeneratedCanonicalCasesMatchTypeScript(t *testing.T) {
	raw, err := os.ReadFile("fixtures/canonical/generated-cases.json")
	if err != nil {
		t.Fatal(err)
	}
	var fixture struct {
		Algorithm        string
		BoundedCaseCount int
		Cases            []struct {
			Name      string
			Seed      string
			Input     json.RawMessage
			Canonical string
			SHA256    string
		}
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatal(err)
	}
	if fixture.Algorithm != "lcg32-v1" {
		t.Fatalf("unexpected generated fixture algorithm %q", fixture.Algorithm)
	}
	if len(fixture.Cases) == 0 {
		t.Fatal("generated fixture has no canonical cases")
	}
	if len(fixture.Cases) != fixture.BoundedCaseCount {
		t.Fatalf("generated case count %d, want %d", len(fixture.Cases), fixture.BoundedCaseCount)
	}
	for _, c := range fixture.Cases {
		got, err := CanonicalJSONBytes(c.Input)
		if err != nil {
			t.Errorf("%s seed=%s: %v", c.Name, c.Seed, err)
			continue
		}
		if got != c.Canonical {
			t.Errorf("%s seed=%s:\n got %s\nwant %s", c.Name, c.Seed, got, c.Canonical)
		}
		if SHA256Hex(got) != c.SHA256 {
			t.Errorf("%s seed=%s: sha256 mismatch", c.Name, c.Seed)
		}
	}
}

func TestCanonicalRejectsNonFinite(t *testing.T) {
	if _, err := CanonicalJSON(map[string]any{"a": []any{1.0, mathInf()}}); err == nil || !strings.Contains(err.Error(), "$.a[1]") {
		t.Fatalf("expected non-finite error with path, got %v", err)
	}
	got, _ := CanonicalJSON(map[string]any{"b": negZero()})
	if got != `{"b":0}` {
		t.Fatalf("-0 must canonicalise to 0, got %s", got)
	}
}

func TestGeneratedFingerprintsMatchTypeScript(t *testing.T) {
	raw, err := os.ReadFile("fixtures/canonical/generated-cases.json")
	if err != nil {
		t.Fatal(err)
	}
	var fixture struct {
		Fingerprints []struct {
			Name            string
			Seed            string
			Manifest        json.RawMessage
			IdentityVariant map[string]any
			MaterialVariant struct {
				Costs map[string]any
			}
			Fingerprint         string
			MaterialFingerprint string
		}
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatal(err)
	}
	if len(fixture.Fingerprints) == 0 {
		t.Fatal("generated fixture has no fingerprint cases")
	}
	for _, c := range fixture.Fingerprints {
		manifest := decodeJSONMap(t, c.Manifest)
		got, err := InputFingerprint(manifest)
		if err != nil {
			t.Errorf("%s seed=%s: %v", c.Name, c.Seed, err)
			continue
		}
		if got != c.Fingerprint {
			t.Errorf("%s seed=%s: fingerprint %s, want %s", c.Name, c.Seed, got, c.Fingerprint)
		}
		identity := cloneJSONMap(t, manifest)
		for key, value := range c.IdentityVariant {
			identity[key] = value
		}
		again, err := InputFingerprint(identity)
		if err != nil || again != got {
			t.Errorf("%s seed=%s: identity changed fingerprint to %s (%v)", c.Name, c.Seed, again, err)
		}
		material := cloneJSONMap(t, manifest)
		costs, ok := material["costs"].(map[string]any)
		if !ok {
			t.Fatalf("%s seed=%s: manifest costs is %T", c.Name, c.Seed, material["costs"])
		}
		for key, value := range c.MaterialVariant.Costs {
			costs[key] = value
		}
		changed, err := InputFingerprint(material)
		if err != nil || changed != c.MaterialFingerprint {
			t.Errorf("%s seed=%s: material fingerprint %s, want %s (%v)", c.Name, c.Seed, changed, c.MaterialFingerprint, err)
		}
		if changed == got {
			t.Errorf("%s seed=%s: material change kept fingerprint %s", c.Name, c.Seed, changed)
		}
	}
}

func TestInputFingerprintMatchesTypeScript(t *testing.T) {
	raw, err := os.ReadFile("fixtures/validation-run-manifest.v1/valid/example.json")
	if err != nil {
		t.Fatal(err)
	}
	want, err := os.ReadFile("fixtures/validation-run-manifest.v1/fingerprint.txt")
	if err != nil {
		t.Fatal(err)
	}
	value, err := Validate("heisentick/validation-run-manifest", raw)
	if err != nil {
		t.Fatal(err)
	}
	manifest := value.(map[string]any)
	got, err := InputFingerprint(manifest)
	if err != nil {
		t.Fatal(err)
	}
	if got != strings.TrimSpace(string(want)) {
		t.Fatalf("fingerprint %s, TypeScript wrote %s", got, strings.TrimSpace(string(want)))
	}
	manifest["runId"] = "9f3c1d2e-4a5b-4c6d-8e9f-0a1b2c3d4e5f"
	manifest["submittedAt"] = json.Number("1")
	again, _ := InputFingerprint(manifest)
	if again != got {
		t.Fatal("identity fields must not change the fingerprint")
	}
	costs := manifest["costs"].(map[string]any)
	costs["slippage"] = json.Number("0.07")
	changed, _ := InputFingerprint(manifest)
	if changed == got {
		t.Fatal("an input change must change the fingerprint")
	}
}

func decodeJSONMap(t *testing.T, raw json.RawMessage) map[string]any {
	t.Helper()
	dec := json.NewDecoder(strings.NewReader(string(raw)))
	dec.UseNumber()
	var value map[string]any
	if err := dec.Decode(&value); err != nil {
		t.Fatal(err)
	}
	return value
}

func cloneJSONMap(t *testing.T, value map[string]any) map[string]any {
	t.Helper()
	raw, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	return decodeJSONMap(t, raw)
}
